geoarrow-types/src/geoarrow/types/type_base.py

```diff
@@ -0,0 +1,184 @@
+from typing import Optional
+import json
+from geoarrow.types.crs import Crs
+from geoarrow.types.constants import (
+    Encoding,
+    GeometryType,
+    Dimensions,
+    CoordType,
+    EdgeType,
+)
+
+
+class GeoArrowType:
+    def __init__(self) -> None:
+        self._encoding = Encoding.UNKNOWN
+        self._geometry_type = GeometryType.GEOMETRY
+        self._dimensions = Dimensions.UNKNOWN
+        self._coord_type = CoordType.UNKNOWN
+        self._edge_type = EdgeType.UNKNOWN
+        self._crs = None
+
+    def _check_crs(self, crs):
+        if crs is None or hasattr(crs, "to_json_dict"):
+            return crs
+
+        raise TypeError("crs must be None or have a to_json_dict()  method")
+
+    @property
+    def encoding(self) -> Encoding:
+        return self._encoding
+
+    @property
+    def geometry_type(self) -> GeometryType:
+        return self._geometry_type
+
+    @property
+    def dimensions(self) -> Dimensions:
+        return self._dimensions
+
+    @property
+    def coord_type(self) -> CoordType:
+        return self._coord_type
+
+    @property
+    def edge_type(self) -> EdgeType:
+        return self._edge_type
+
+    @property
+    def crs(self) -> Optional[Crs]:
+        return self._crs
+
+    @property
+    def extension_metadata(self) -> str:
+        metadata = {}
+
+        if self.crs is not None:
+            metadata["crs"] = self.crs.to_json_dict()
+
+        if self.edge_type == EdgeType.SPHERICAL:
+            metadata["edges"] = "spherical"
+
+        return json.dumps(metadata)
+
+    @property
+    def extension_name(self) -> str:
+        raise NotImplementedError()
+
+    def with_encoding(self, encoding: Encoding) -> "SerializedType":
+        raise NotImplementedError()
```

Comment:
This doesn't return `Self`?

Reply:
You can use `typing_extensions` for pre 3.11

Reply:
But regardless of whether you use `Self`, my point here was that this method on `GeoArrowType` does not return a `GeoArrowType`, but rather a `SerializedType`. Given that `GEOARROW` is a variant of the `Encoding` enum, it seems weird that `GeoArrowType.with_encoding(Encoding.GEOARROW)` would return a `SerializedType`.